website/integrations: headscale (#14500)
goauthentik/authentik: website/integrations/services/headscale/index.md

---
title: Integrate with Headscale
sidebar_label: Headscale
support_level: community
---

## What is Headscale

> > [Headscale](https://github.com/juanfont/headscale) is an open source alternative to the Tailscale coordination server and can be self-hosted for a single tailnet. Headscale is a re-implemented version of the Tailscale coordination server, developed independently and completely separate from Tailscale.

Reviewer: ```suggestion
> [Headscale](https://github.com/juanfont/headscale) is an open source alternative to the Tailscale coordination server and can be self-hosted for a single tailnet. Headscale is a re-implemented version of the Tailscale coordination server, developed independently and completely separate from Tailscale, with its own independent community of users and developers.
```

> Headscale is a project that complements Tailscale — with its own independent community of users and developers.

Reviewer: ```suggestion
```

---
title: Integrate with Headscale
sidebar_label: Headscale
support_level: community
---

## What is Headscale

> > Headscale is an open-source, self-hosted server that acts as a Tailscale control server.

Reviewer: ```suggestion
> [Headscale](https://github.com/juanfont/headscale) is an open source alternative to the Tailscale coordination server and can be self-hosted for a single tailnet. Headscale is a re-implemented version of the Tailscale coordination server, developed independently and completely separate from Tailscale.
Headscale is a project that complements Tailscale — with its own independent community of users and developers.
```

Reviewer: Let's take the full, longer description from their site. There are more details than can fit into a single sentence. :-)

:::

## authentik configuration

To support the integration of Headscale with authentik, you need to create an application/provider pair in authentik.

### Create an application and provider in authentik

1. Log in to authentik as an administrator and open the authentik Admin interface.
2. Navigate to **Applications** > **Applications** and click **Create with Provider** to create an application and provider pair. (Alternatively you can first create a provider separately, then create the application and connect it with the provider.)

- **Application**: provide a descriptive name, an optional group for the type of application, the policy engine mode, and optional UI settings.
- **Choose a Provider type**: select **OAuth2/OpenID Connect** as the provider type.
- **Configure the Provider**: provide a name (or accept the auto-provided name), the authorization flow to use for this provider, and the following required configurations.
    - Note the **Client ID**,**Client Secret**, and **slug** values because they will be required later.
    - Set a `Strict` redirect URI to `https://headscale.company/oidc/callback`.
    - Select any available signing key.

> - **Configure Bindings** _(optional)_: you can create a [binding](/docs/add-secure-apps/flows-stages/bindings/) (policy, group, or user) to manage the listing and access to applications on a user's **My applications** page.

Reviewer 2: These lines should be indented

3. Click **Submit** to save the new application and provider.

## Headscale configuration

> To support the integration of Headscale with authentik, you'll need to update your `config.yaml`:

Reviewer 2: ```suggestion
To support the integration of Headscale with authentik, you'll need to update the `config.yaml` of your headscale deployment:
```

    # client_secret_path: "${CREDENTIALS_DIRECTORY}/oidc_client_secret"
    #
    # 3. From an environment variable:
    # The HEADSCALE_OIDC_CLIENT_SECRET environment variable will be used automatically if set

    # OIDC scopes to request (defaults: "openid", "profile", "email")
    # Additional scopes can be added to request extra user information
    scope: ["openid", "profile", "email", "custom"]
    # Passed on to the browser login request – used to tweak behaviour for the OIDC provider (optional)
    extra_params:
        domain_hint: acmecorp.net

    # Reject authentication if user doesn't match these criteria (optional)
    allowed_domains:
        - acmecorp.net

    # Group-based access control (optional)
    allowed_groups:
        - /headscale

    # Specific user allowlist (optional)
    allowed_users:
        - dominic@acmecorp.net

    pkce:
        enabled: true
        method: S256

    # Email Domain Handling:
    # When strip_email_domain is true: "user@acmecorp.net" becomes username "user"
    # When strip_email_domain is false: "user@acmecorp.net" becomes username "user.acmecorp.net"
    strip_email_domain: true
```

> ## Resources

Reviewer 2: Can you move the resources section to the end

- [Headscale OIDC documentation](https://headscale.net/stable/ref/oidc/#basic-configuration)

## Configuration verification

> To verify the integration with Headscale, log out and attempt to log back in using OIDC.

Author: todo: validate
